src/jit/gcinfo.cpp

```diff
@@ -524,56 +463,128 @@ void GCInfo::gcCountForHeader(UNALIGNED unsigned int* untrackedCount, UNALIGNED
         }
 #endif
 
-        count++;
+        untrackedCount++;
     }
 
 #ifdef DEBUG
     if (compiler->verbose)
     {
-        printf("GCINFO: untrckVars = %u\n", count);
+        printf("GCINFO: untrckVars = %u\n", untrackedCount);
     }
 #endif
 
-    *untrackedCount = count;
+    *pUntrackedCount = untrackedCount;
 
-    /* Count the number of entries in the table of non-register pointer
-       variable lifetimes. */
+    // Count the number of entries in the table of non-register pointer variable lifetimes.
 
-    count = 0;
+    unsigned int varPtrTableSize = 0;
 
     if (thisKeptAliveIsInUntracked)
     {
-        count++;
+        varPtrTableSize++;
     }
 
-    if (gcVarPtrList)
+    if (gcVarPtrList != nullptr)
     {
-        /* We'll use a delta encoding for the lifetime offsets */
+        // We'll use a delta encoding for the lifetime offsets.
 
-        for (varTmp = gcVarPtrList; varTmp; varTmp = varTmp->vpdNext)
+        for (varPtrDsc* varTmp = gcVarPtrList; varTmp != nullptr; varTmp = varTmp->vpdNext)
         {
-            /* Special case: skip any 0-length lifetimes */
+            // Special case: skip any 0-length lifetimes.
 
             if (varTmp->vpdBegOfs == varTmp->vpdEndOfs)
             {
                 continue;
             }
 
-            count++;
+            varPtrTableSize++;
         }
     }
 
 #ifdef DEBUG
     if (compiler->verbose)
     {
-        printf("GCINFO: trackdLcls = %u\n", count);
+        printf("GCINFO: trackdLcls = %u\n", varPtrTableSize);
     }
 #endif
 
-    *varPtrTableSize = count;
+    *pVarPtrTableSize = varPtrTableSize;
+}
+
+//------------------------------------------------------------------------
+// gcIsUntrackedLocalOrNonEnregisteredArg: Check if this varNum with GC type
+// corresponds to an untracked local or argument that was not fully enregistered.
+//
+//
+// Arguments:
+//   varNum - the variable number to check;
+//   pThisKeptAliveIsInUntracked - if !WIN64EXCEPTIONS and the argument != nullptr remember if `this` should be
+//   untracked and kept alive.
+//
+// Return value:
+//   true if it an untracked pointer value.
+//
+bool GCInfo::gcIsUntrackedLocalOrNonEnregisteredArg(unsigned varNum, bool* pThisKeptAliveIsInUntracked)
```

Comment:
The naming and comments around `pThisKeptAliveIsInUntracked` are kind of confusing. I'd probably rename this to simply `pIsKeepAliveThis`. In fact, it seems that this is set to true in the case where it is going to be tracked (as the untracked area doesn't support reporting of this). If that's the case, I think this needs more clarification.

Reply:
renamed to `keepThisAlive`. 

I agree that this code is confusing and because there are so many code paths there and many of them are not tested (for example x86 Linux, where `jit32_gcencode == true, FEATURE_EH_FUNCLETS == true`) and the biggest part of it is very old (for example, `void emitter::emitGCregLiveSet(GCtype gcType, regMaskTP regMask, BYTE* addr, bool isThis)` has `//  assert(emitFullyInt || isThis);` that was commented out before 2010 and we do not have a history why).

So it is very hard to make any semantic changes there, this PR is cleaning and should not change the existing behavior.

> but will have to be reported as tracked if we have lvaKeepAliveAndReportThis(). Is that the case?

It looks like, but after spending a few hours trying to cover different cases I am still not absolutely sure.

Reply:
OK, thanks. At least now the logic is separated out and a bit easier to discern. It may actually be worth it to at least add a note that the semantics here aren't entirely clear, so that someone looking at this in the future will at least not be surprised at the apparent contradictions here.